### PR TITLE
new and improved flow conditioning

### PR DESCRIPTION
**Pull request by JCBrouwer**
- flow conditioning now agnostic to video resolution
- a 3rd flow edge detection channel was added so that the conditioning can be written to RGB video

### CODE REVIEW

File: prepare_conditioning_dataset.py [hunk 1]
```diff
@@ -3,36 +3,70 @@
...
 from torch.utils.data import DataLoader, Dataset
-from torchvision.io import read_video
+from torchvision.io import read_video, write_video
 from torchvision.models.optical_flow import Raft_Large_Weights, raft_large
 from tqdm import tqdm
 
 
 class FlowPreprocessor(nn.Module):
-    out_channels: int = 2
+    out_channels: int = 3
```
Comment (jfischoff) [inline prepare_conditioning_dataset.py:16]: I'm not really sure how this is used, but doesn't the flow only have two channels?

Comment (JCBrouwer) [reply prepare_conditioning_dataset.py:16]: Yeah, indeed, but I spent a ton of time struggling with video pixel formats to see if there's any way to save a video with only 2 color channels. Turns out to be at least very hard if not impossible.

Instead I've just opted to add a 3rd channel which represents the flow edges (around occlusions). This allows us to save videos and maybe makes the conditioning information a bit more informative.

File: prepare_conditioning_dataset.py [hunk 2]
```diff
@@ -3,36 +3,70 @@
...
-        video = video.mul(2).sub(1)
-        return self.raft.forward(video[:-1], video[1:])[-1]
+        # prepare and run RAFT
+        prev, curr = video[:-1], video[1:]
+        prev, curr = self.transform(prev, curr)
+        flow = self.raft.forward(prev, curr)[-1]
+
+        # smooth flow
+        sigma = min(flow.shape[3], flow.shape[2]) * self.smooth
+        flow = gaussian_blur2d(flow, kernel_size=5, sigma=(sigma, sigma))
```
Comment (jfischoff) [inline prepare_conditioning_dataset.py:34]: Why smooth it?

File: prepare_conditioning_dataset.py [hunk 3]
```diff
@@ -3,36 +3,70 @@
...
+
+        # calculate flow laplacian magnitude (rough edge detection)
+        flow_laplacian = laplacian(flow, kernel_size=5)
+        magnitude = flow_laplacian.abs().sum(1, keepdim=True).pow(1 / self.emphasis).mul(2).sub(1)
+
+        # concatenate to 3-channel conditioning
+        conditioning = torch.cat((flow, magnitude), dim=1)
+
+        # emphasize small values
+        conditioning = torch.sign(conditioning) * conditioning.abs().pow(1 / self.emphasis)
```
Comment (jfischoff) [inline prepare_conditioning_dataset.py:51]: Interesting. I've done similar things with optical flow before as well but I would assume that since we are training with it, this will happen some what automatically?

**Review by jfischoff — APPROVED** [review]
Looks fine, but I think we should try different tests with different preprocessing of the flows to see what works.

File: prepare_conditioning_dataset.py (hunk 2 above)
Comment (JCBrouwer) [reply prepare_conditioning_dataset.py:34]: So we don't get edge detection on the RAFT grid-like artifacts

File: prepare_conditioning_dataset.py (hunk 3 above)
Comment (JCBrouwer) [reply prepare_conditioning_dataset.py:51]: Presumably it will, but a little extra normalization upfront never hurt anybody

**Comment by JCBrouwer** [comment]
Yeah I think those tests would be interesting, but at the moment I don't know if we have very strong metrics to compare. We'll need to think a bit about evaluation in general...
